File: src/jp/yumyum/ShowArea.java
package jp.yumyum;

import android.content.Context;
import android.graphics.Canvas;
import android.view.MotionEvent;
import android.widget.HorizontalScrollView;

class ShowArea extends HorizontalScrollView {
	private GraphView graphView;
	private ValueView vv;

	public ShowArea(Context context, ValueView vview) {
		super(context);
		this.setHorizontalFadingEdgeEnabled(false);
		vv = vview;
	}

	@Override
	protected void onLayout(boolean changed, int l, int t, int r, int b) {

		// 自動スクロール中なら常にGraphViewの右端を表示
		if (graphView != null && graphView.isScrolling()) {
			scrollTo(graphView.getWidth() - this.getWidth(), 0);
		}
		super.onLayout(changed, l, t, r, b);
	}

	@Override
	public boolean onTouchEvent(MotionEvent ev) {
		// 自動スクロール中ならタッチされても何もしない
		if (graphView.isScrolling()) {
			return true;
		}
		return super.onTouchEvent(ev);
	}

	@Override
	protected void onDraw(Canvas canvas) {
		// graphViewが空なら（初めて呼び出されるなら）作成して、子供Viewとして登録
		if (graphView == null) {
			graphView = new GraphView(this.getContext(), this.getWidth(),
					this.getHeight(), vv);
			this.addView(graphView);
			invalidate();
		}

		super.onDraw(canvas);
	}

	public void tapEvent(long time) {
		if (graphView.isScrolling() == false) {
			// 自動スクロール中はスクロールバー非表示
			this.setHorizontalScrollBarEnabled(false);
		}
		graphView.tap(time);
		return;
	}

	public void reset() {
		graphView.reset();
	}

	public int getTagetBpm() {
		return graphView.getTargetBpm();
	}

	public void stopScroll() {
		if (graphView != null)
			graphView.stopScroll();
	}

}

File: app/src/main/java/jp/yumyum/ShowArea.kt
package jp.yumyum

import android.content.Context
import android.graphics.Canvas
import android.view.MotionEvent
import android.widget.HorizontalScrollView

internal class ShowArea(context: Context, private val vv: ValueView) : HorizontalScrollView(context) {
    private var graphView: GraphView? = null

//    val tagetBpm: Int
//        get() = graphView!!.targetBpm
//
//    init {
//        this.isHorizontalFadingEdgeEnabled = false
//    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {

        // 自動スクロール中なら常にGraphViewの右端を表示
        if (graphView != null && graphView!!.isScrolling) {
            scrollTo(graphView!!.width - this.width, 0)
        }
        super.onLayout(changed, l, t, r, b)
    }

    override fun onTouchEvent(ev: MotionEvent): Boolean {
        // 自動スクロール中ならタッチされても何もしない
        return if (graphView!!.isScrolling) {
            true
        } else super.onTouchEvent(ev)
    }

    override fun onDraw(canvas: Canvas) {
        // graphViewが空なら（初めて呼び出されるなら）作成して、子供Viewとして登録
        if (graphView == null) {
            graphView = GraphView(this.context, this.width,
                    this.height, vv)
            this.addView(graphView)
            invalidate()
        }
        super.onDraw(canvas)
    }

    fun tapEvent(time: Long) {
        if (!graphView!!.isScrolling) {
            // 自動スクロール中はスクロールバー非表示
            this.isHorizontalScrollBarEnabled = false
        }
        graphView!!.tap(time)
        return
    }

    fun reset() {
        graphView!!.reset()
    }

    fun stopScroll() {
        if (graphView != null)
            graphView!!.stopScroll()
    }

}
